crates/ra_hir/src/ty/tests.rs

```diff
@@ -586,6 +586,139 @@ fn test() -> i128 {
     );
 }
 
+#[test]
+fn infer_associated_const() {
+    check_inference(
+        "infer_associated_const",
+        r#"
+struct Struct;
+
+impl Struct {
+    const FOO: u32 = 1;
+}
+
+enum Enum;
+
+impl Enum {
+    const BAR: u32 = 2;
+}
+
+trait Trait {
+    const ID: u32;
+}
+
+struct TraitTest;
+
+impl Trait for TraitTest {
+    const ID: u32 = 5;
+}
+
+fn test() {
+    let x = Struct::FOO;
+    let y = Enum::BAR;
+    let z = TraitTest::ID;
+}
+"#,
+    );
+}
+
+#[test]
+fn infer_associated_method_struct() {
+    check_inference(
+        "infer_associated_method_struct",
+        r#"
+struct A { x: u32 };
+
+impl A {
+    fn new() -> A {
+        A { x: 0 }
+    }
+}
+fn test() {
+    let a = A::new();
+    a.x;
+}
+"#,
+    );
+}
+
+#[test]
+fn infer_associated_method_enum() {
+    check_inference(
+        "infer_associated_method_enum",
+        r#"
+enum A { B, C };
+
+impl A {
+    pub fn b() -> A {
+        A::B
```

Comment:
it's not really necessary to fill out these functions in the test -- only the signature matters